index.js

```diff
@@ -78,85 +69,97 @@ function formy(ctx, opts) {
  * @see https://github.com/dlau/koa-body
  * @api public
  */
-function requestbody(opts) {
-  const mergedOpts = opts || {};
-  mergedOpts.onError = 'onError' in mergedOpts ? mergedOpts.onError : false;
-  mergedOpts.multipart = 'multipart' in mergedOpts ? mergedOpts.multipart : false;
-  mergedOpts.urlencoded = 'urlencoded' in mergedOpts ? mergedOpts.urlencoded : true;
-  mergedOpts.json = 'json' in mergedOpts ? mergedOpts.json : true;
-  mergedOpts.text = 'text' in mergedOpts ? mergedOpts.text : true;
-  mergedOpts.encoding = 'encoding' in mergedOpts ? mergedOpts.encoding : 'utf-8';
-  mergedOpts.jsonLimit = 'jsonLimit' in mergedOpts ? mergedOpts.jsonLimit : '1mb';
-  mergedOpts.jsonStrict = 'jsonStrict' in mergedOpts ? mergedOpts.jsonStrict : true;
-  mergedOpts.formLimit = 'formLimit' in mergedOpts ? mergedOpts.formLimit : '56kb';
-  mergedOpts.queryString = 'queryString' in mergedOpts ? mergedOpts.queryString : null;
-  mergedOpts.formidable = 'formidable' in mergedOpts ? mergedOpts.formidable : {};
-  mergedOpts.includeUnparsed = 'includeUnparsed' in mergedOpts ? mergedOpts.includeUnparsed : false;
-  mergedOpts.textLimit = 'textLimit' in mergedOpts ? mergedOpts.textLimit : '56kb';
-
-  mergedOpts.parsedMethods = 'parsedMethods' in mergedOpts ? mergedOpts.parsedMethods : ['POST', 'PUT', 'PATCH'];
-  mergedOpts.parsedMethods = mergedOpts.parsedMethods.map(method => method.toUpperCase());
-
-  return function closure(ctx, next) {
-    let bodyPromise;
+function requestbody(_opts) {
+  const opts = _opts || {};
+
+  opts.onError = 'onError' in opts ? opts.onError : ((err) => { throw err; });
```

Comment:
I'm not a fan of defining this error closure here because it'll produce odd stacktraces compared to a simpler (to reason about) if/else statement in the try/catch block below.

Reply:
Is there a reason why we'd default it here that I'm not seeing the benefit of?

Reply:
Mostly the uniform throw case when not passing a function as onError.  This was before I had a single try/catch too, though, which has probably made it a moot point there.

I'm *far* more concerned though with getting some uniform handling of request body errors.  The two options I presented in the PR summary of always throwing or mandatory throwing in the middleware are what I'm stuck on, and haven't come to a concrete resolution between.

Reply:
I've given the two options some thought and I've decided that I don't want to suppress errors users choose to throw or not throw from the `onError` function. In my mind, suppressing a callback error and handling throwing (even if we rethrow) would be an unexpected library behaviour.

If the user wants to suppress or even throw a string, we should let them.

Reply:
That being said, failure to parse a body is technically an HTTP 400 by definition.  The application cannot be expected to operate correctly if the precondition of a correctly-parsed body is not met, and it's more than likely a request level issue.

Being able to work around it is nice, but it's easily a footgun we're handing the user.

Reply:
I agree with all your points; I don't like `onError` either. Maybe we add to the docs that it's not recommended for use and to handle things in middleware?

Reply:
If you have a feature that you don't recommend, isn't that the essence of deprecation?

Why support functionality you have to advise against using?

Reply:
I suppose only so that you don't break even more things. 😉 

`onError` originated here: https://github.com/dlau/koa-body/pull/41

I'm OK with removing it.

Reply:
Yeah, looking at the original PR I don't see any great reasons provided for implementing it at the time even.  Looking at `koa-bodyparser` as well, I see no real justification there for when they implemented it.

Reply:
@damianb Sorry to bounce back to this, but I think I've come up with a reason for `onError` after looking at our codebase, but maybe there's a better approach...

We use `onError` to throw [Boom errors](https://github.com/hapijs/boom/blob/master/API.md#boombadrequestmessage-data) from koa-body when parsing fails. I suppose an alternative approach would be to provide a wrapper around koa-body.

Reply:
Hmm. Yeah, because deeper libraries are just going to throw an error, so unless you inspect the error itself you're not able to mutate it into something useful.

In which case...I think we can take a step back and put onError back in, but we *have* to mandate a throw.  If the onError doesn't throw, koa-body should throw immediately after.